# SQL

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
query = '''
SELECT * FROM books LIMIT 5;
'''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<div style="background-color: #F2F8FD; padding: 10px;">
    <b>Таблица `books`</b>

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.</div>

In [4]:
query = '''
SELECT * FROM authors LIMIT 5;
'''
pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<div style="background-color: #F2F8FD; padding: 10px;">
    <b>Таблица `authors`</b>

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.</div>

In [5]:
query = '''
SELECT * FROM publishers LIMIT 5;
'''
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<div style="background-color: #F2F8FD; padding: 10px;">
    <b>Таблица `publishers`</b>

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;</div>

In [6]:
query = '''
SELECT * FROM ratings LIMIT 5;
'''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<div style="background-color: #F2F8FD; padding: 10px;">
    <b>Таблица `ratings`</b>

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.</div>

In [7]:
query = '''
SELECT * FROM reviews LIMIT 5;
'''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div style="background-color: #F2F8FD; padding: 10px;">
    <b>Таблица `reviews`</b>

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.</div>

### Задания

**1. Посчитайте количество книг, выпущенных после 1 января 2000 года**

In [8]:
# Для этого посчитаем все строки в таблице books в которых publication_date больше '2000-01-01'
query = '''
SELECT 
    COUNT(*)
FROM
    books
WHERE 
    publication_date > '2000-01-01';
'''
pd.io.sql.read_sql(query, con = engine)

count
0    819

<div style="background-color: #FFFCBB; padding: 10px;">
    После 1 января 2000 года выпущено 819 книг.
</div>

**2. Посчитайте количество пользовательских обзоров и среднюю оценку для каждой книги**

In [9]:
query = '''
SELECT -- выведем название книги, средний рейтинг и количество обзоров
    books.title AS title,
    AVG(ratings.rating) AS avg_rating,
    COUNT(reviews.review_id) AS cnt_review_id
FROM -- к таблице books добавим табилцы ratings и reviews по полю book_id
    books
LEFT OUTER JOIN ratings ON ratings.book_id = books.book_id
LEFT OUTER JOIN reviews ON reviews.book_id = books.book_id
GROUP BY 
    books.title
ORDER BY -- отсортируем получившуюся таблицу по количеству обзоров по убыванию
    cnt_review_id DESC;
'''
pd.io.sql.read_sql(query, con = engine)

title  avg_rating  \
0                              Twilight (Twilight  #1)    3.662500   
1                  The Hobbit  or There and Back Again    4.125000   
2                               The Catcher in the Rye    3.825581   
3    Harry Potter and the Prisoner of Azkaban (Harr...    4.414634   
4    Harry Potter and the Chamber of Secrets (Harry...    4.287500   
..                                                 ...         ...   
994                            The Natural Way to Draw    3.000000   
995                          Essential Tales and Poems    4.000000   
996   The Cat in the Hat and Other Dr. Seuss Favorites    5.000000   
997                               Leonardo's Notebooks    4.000000   
998  Disney's Beauty and the Beast (A Little Golden...    4.000000   

     cnt_review_id  
0             1120  
1              528  
2              516  
3              492  
4              480  
..             ...  
994              0  
995              0  
996              0  
997              0  
998              0  

[999 rows x 3 columns]

<div style="background-color: #FFFCBB; padding: 10px;">
    Больше всего обзоров пользователи написали на Twilight, правда, средний рейтинг у него не очень высок.
</div>

**3. Определите издательство, которое издало наибольшее число книг толще 50 страниц (так вы исключите из анализа различные брошюры)**

In [10]:
query = '''
SELECT -- выведем название издательства и количество книг
    publishers.publisher AS publisher,
    COUNT(books.book_id) AS cnt_books
FROM -- к таблице books добавим табилцу publishers по полю publisher_id
    books
LEFT OUTER JOIN publishers ON publishers.publisher_id = books.publisher_id
WHERE -- зададим фильтр по количеству страниц в книге больше 50
    num_pages > '50'
GROUP BY
    publisher
ORDER BY -- отсортируем получившуюся таблицу по количеству книг по убыванию
    cnt_books DESC
LIMIT 1; -- напечатаем первую строку таблицы
'''
pd.io.sql.read_sql(query, con = engine)

publisher  cnt_books
0  Penguin Books         42

<div style="background-color: #FFFCBB; padding: 10px;">
    Издательство, которое издало наибольшее число книг (а конкретнее, 42 книги) толще 50 страниц - Penguin Books.
</div>

**4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более пользовательскими оценками**

In [11]:
query = '''
SELECT
    author -- выбираем автора
FROM
    authors -- из таблицы авторов
WHERE -- с айди полученным в подзапросе
    author_id =
        (SELECT -- выбираем айди автора
            author_id
        FROM -- из подзапроса
            (SELECT -- выбираем author_id и считаем средний рейтинг по автору
                books.author_id,
                AVG(ratings.rating) AS avg_rating
            FROM --из таблицы ratings
                ratings
            LEFT JOIN books ON books.book_id = ratings.book_id --добавив author_id из таблицы books
            WHERE -- где book_id из подзапроса
                ratings.book_id IN
                    (SELECT -- выбираем book_id
                         book_id
                    FROM -- из таблицы ratings
                         ratings
                    GROUP BY -- группируем по book_id
                        book_id
                    HAVING -- выбираем только книги с 50 и более оценками
                        COUNT(ratings.rating_id) >= '50')
            GROUP BY -- группируем по author_id
                author_id
            ORDER BY -- сортируем по среднему рейтингу по убыванию
                avg_rating DESC
            LIMIT 1) AS Sub);
                    

'''
pd.io.sql.read_sql(query, con = engine)

author
0  J.K. Rowling/Mary GrandPré

<div style="background-color: #FFFCBB; padding: 10px;">
    Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré.<br>
</div>

**5. Посчитайте среднее количество текстовых обзоров пользователей, которые поставили более чем по 50 оценок.**

In [12]:
query = '''
SELECT -- выведем среднее по количеству обзоров полученных в подзапросе Sub
    AVG(Sub.cnt_reviews) AS avg_reviews
FROM
    (SELECT  -- выбираем username и считаем количество обзоров
        reviews.username,
        COUNT(reviews.review_id) AS cnt_reviews
    FROM -- из таблицы с обзорами
        reviews
    WHERE -- по username, которые получили в подзапросе
        username IN
            (SELECT -- выбираем username
                ratings.username
            FROM -- из таблицы с оценками
                ratings
            GROUP BY
                ratings.username
            HAVING -- которые поставили более, чем по 50 оценок
                COUNT(ratings.rating_id) > '50')
    GROUP BY
        reviews.username) AS Sub;
'''
pd.io.sql.read_sql(query, con = engine)

avg_reviews
0    24.333333

<div style="background-color: #FFFCBB; padding: 10px;">
    Среднее количество текстовых обзоров пользователей, которые поставили более чем по 50 оценок - 24.33.<br>
</div>